fastly/block_fastly_service_v1_s3logging.go

```diff
@@ -135,84 +159,103 @@ func processS3Logging(d *schema.ResourceData, conn *gofastly.Client, latestVersi
-			Service: d.Id(),
-			Version: latestVersion,
-			Name:    sf["name"].(string),
-		}
+		opts := buildDeleteS3(sf, serviceID, latestVersion)
 
 		log.Printf("[DEBUG] Fastly S3 Logging removal opts: %#v", opts)
-		err := conn.DeleteS3(&opts)
-		if errRes, ok := err.(*gofastly.HTTPError); ok {
-			if errRes.StatusCode != 404 {
-				return err
-			}
-		} else if err != nil {
+
+		if err := deleteS3(conn, opts); err != nil {
 			return err
 		}
 	}
 
-	// POST new/updated S3 Logging
+	// POST new/updated S3 Logging.
 	for _, sRaw := range addS3Logging {
 		sf := sRaw.(map[string]interface{})
 
-		// Fastly API will not error if these are omitted, so we throw an error
-		// if any of these are empty
-		for _, sk := range []string{"s3_access_key", "s3_secret_key"} {
-			if sf[sk].(string) == "" {
-				return fmt.Errorf("[ERR] No %s found for S3 Log stream setup for Service (%s)", sk, d.Id())
-			}
+		opts, err := buildCreateS3(sf, serviceID, latestVersion)
+		if err != nil {
+			return err
 		}
 
-		opts := gofastly.CreateS3Input{
-			Service:                      d.Id(),
-			Version:                      latestVersion,
-			Name:                         sf["name"].(string),
-			BucketName:                   sf["bucket_name"].(string),
-			AccessKey:                    sf["s3_access_key"].(string),
-			SecretKey:                    sf["s3_secret_key"].(string),
-			Period:                       uint(sf["period"].(int)),
-			GzipLevel:                    uint(sf["gzip_level"].(int)),
-			Domain:                       sf["domain"].(string),
-			Path:                         sf["path"].(string),
-			Format:                       sf["format"].(string),
-			FormatVersion:                uint(sf["format_version"].(int)),
-			TimestampFormat:              sf["timestamp_format"].(string),
-			ResponseCondition:            sf["response_condition"].(string),
-			MessageType:                  sf["message_type"].(string),
-			Placement:                    sf["placement"].(string),
-			ServerSideEncryptionKMSKeyID: sf["server_side_encryption_kms_key_id"].(string),
+		log.Printf("[DEBUG] Create S3 Logging Opts: %#v", opts)
+		_, err := conn.CreateS3(&opts)
+		if err != nil {
+			return err
 		}
+	}
+	return nil
+}
+
+func readS3Logging(conn *gofastly.Client, d *schema.ResourceData, s *gofastly.ServiceDetail) error {
+	log.Printf("[DEBUG] Refreshing S3 Logging for (%s)", d.Id())
+	s3List, err := conn.ListS3s(&gofastly.ListS3sInput{
+		Service: d.Id(),
+		Version: s.ActiveVersion.Number,
+	})
+
+	if err != nil {
+		return fmt.Errorf("[ERR] Error looking up S3 Logging for (%s), version (%v): %s", d.Id(), s.ActiveVersion.Number, err)
+	}
 
-		redundancy := strings.ToLower(sf["redundancy"].(string))
-		switch redundancy {
-		case "standard":
-			opts.Redundancy = gofastly.S3RedundancyStandard
-		case "reduced_redundancy":
-			opts.Redundancy = gofastly.S3RedundancyReduced
+	sl := flattenS3s(s3List)
+
+	if err := d.Set("s3logging", sl); err != nil {
```

Comment:
See this is where state gets set. Therefore, as long as the API always returns a value with a trailing slash, it shouldn't result in a diff. We don't currently have TF validation around this and I would imagine we would have heard about this if it were an issue.